---
title: "Differential expresion and GO terms analysis with HiSat2"
output: pdf_document
params:
  inpath: "../configs"
---

```{r,warning=FALSE,message=FALSE,include=FALSE}

packages<-c("knitr","RColorBrewer","VennDiagram","veccompare","ggplot2","gplots","yaml","gridExtra","statmod")
package.check <- lapply(
  packages,
  FUN = function(x) {
    if (!require(x, character.only = TRUE)) {
      install.packages(x, dependencies = TRUE,repos = "http://cran.us.r-project.org")
      library(x, character.only = TRUE)
    }
  }
)
config<-read_yaml(paste(path_configs,'/config.yaml',sep=''),fileEncoding = "UTF-8")

if (!require("BiocManager", quietly = TRUE))
   { install.packages("BiocManager",repos = "http://cran.us.r-project.org")
BiocManager::install()}
listOfBiocPackages<-c("topGO",config$GO_MAPPING,"biomaRt","limma","DESeq2","edgeR","sva")
notInstalled <- which( !listOfBiocPackages %in% rownames(installed.packages()) )
## check there's still something left to install
if( length(notInstalled) ) {
    BiocManager::install(listOfBiocPackages[ notInstalled ])
}
lapply(listOfBiocPackages,library, character.only=TRUE)
lapply(packages,require, character.only=TRUE)

path_configs<-params$inpath
knitr::opts_chunk$set(echo=FALSE)
hisat_path<-paste(config$FINALOUTPUT,config$PROJECT,'genome/',sep='/')

source("utils.R")
```




# Section 1 - Limma

```{r,warning=FALSE,message=FALSE,include=FALSE}

###################
#Data load
###################
## HiSat2+StringTie
hisat.t <- read.table(file=paste(hisat_path,"transcript_count_matrix.csv",sep='/'), header=TRUE, row.names=1, sep=",")
hisat.g <- read.table(file=paste(hisat_path,"gene_count_matrix.csv",sep='/'), header=TRUE, row.names=1, sep=",")
Description<-read.table(paste(path_configs,'/Description.csv',sep=''),header=TRUE,sep=',', stringsAsFactors=FALSE)

repetitions<-max(Description$Repetition)
if (repetitions <2)
{
  print('Too litlle repetitions, specify at least 2!')
  knitr::knit_exit()
}

```
```{r,warning=FALSE,message=FALSE,include=FALSE}

dataset_size<-dim(Description)[1]
for (i in (1:dataset_size))
{
  Description[i,7]<-paste('B',Description[i,5],'_',Description[i,3],'_',Description[i,4],sep="")
}
colnames(Description)[7]<-c("Name")

colnam2<-c()
for (i in (1:dataset_size))
{
  colnam2[i]<-Description[which(Description$Group==colnames(hisat.g)[i]),7]
}
colnames(hisat.t) <- colnam2
colnames(hisat.g) <- colnam2
hisat.g<-hisat.g[,order(colnames(hisat.g))]
hisat.t<-hisat.t[,order(colnames(hisat.t))]

###################
#Data description
###################

no_of_batches<-length(unique(Description$Batch))
size_of_batches<-c()
for (i in (1:no_of_batches))
{
  size_of_batches<-cbind(size_of_batches,length(which(Description$Batch==i)))
}
if(sum(size_of_batches)!=dim(hisat.g)[2]){
    stop('Sum of all batches sizes does not match input size')
    }
samples<-sort(paste('B',Description$Batch,'_',Description$Condition,sep=''))
names_col<-unique(samples)
group_size<-length(samples)/length(names_col)
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))

######################
#Low values removal
######################

th1 <- 5
th2 <- 2
n1 <- round(dim(hisat.g)[2]/3)
n2 <- n1*2
filter1 <- apply(hisat.g, 1, function(x) length(x[x>th1])>=n1)
filter2 <- apply(hisat.g, 1, function(x) length(x[x>th2])>=n2)
filter <- filter1&filter2

hisat.g.filtered<-hisat.g[filter,]
hisat.g.filtered.log<-log2(hisat.g.filtered+0.5)
 
```

# Heatmap raw data

```{r,warning=FALSE,message=FALSE,fig.align='left'}

IQR<- apply(hisat.g.filtered,1,IQR,na.rm=TRUE)
topVar<-order(IQR, decreasing=TRUE)[1:100]
heatmap.2(as.matrix(hisat.g.filtered.log[topVar,]),
          scale="row",
          hclust=function(x) hclust(x,method="average"),
          distfun=function(x) as.dist((1-cor(t(x)))/2),
          trace="none",
          main= 'Heatmap for 100 most varible samples\n raw data',
          density="none",
          labRow="",
          cexCol=0.5,dendrogram='column',key=FALSE)

```

# PCA raw data

```{r,warning=FALSE,message=FALSE,fig.align='left',fig.height=7,fig.width=7}

############
#PCA before
############
Plot_PCA(hisat.g.filtered.log,samples,names_col,group_size,no_of_batches,size_of_batches,col_vector,"PCA before SVA")
#legend('topleft',legend=names_col,pch=rep(unique(symbols2),lwd=4,size_of_batches/group_size),col=unique(cols2),cex=0.7)


```
# Violinplot raw data

```{r,warning=FALSE,message=FALSE,fig.height = 5}
Plot_violin(hisat.g.filtered,samples,group_size)
```

```{r,warning=FALSE,message=FALSE,include=FALSE}

########
##limma
########

dge <- DGEList(counts=hisat.g.filtered, group=samples)
dge <- calcNormFactors(dge)
## voom transformation
mod1 <- model.matrix(~0+as.factor(samples))
mod0 <- model.matrix(~1,data=as.factor(samples))
colnames(mod1) <- levels(factor(samples))
rownames(mod1) <- colnames(hisat.g.filtered)
v <- voom(dge,mod1,plot=FALSE)
to_sva <- 2^v$E
nsv<-num.sv(as.matrix(hisat.g.filtered),mod1,method="leek") 
an.error.occured <- TRUE
while(an.error.occured==TRUE){
   svseq <- try(svaseq(to_sva,mod1,mod0,const=0,n.sv=nsv),TRUE)
   if(isTRUE(class(svseq)=="try-error")) { 
     nsv <-nsv-1
     next
      } 
   else { an.error.occured <- FALSE}
}
newV=NULL
fsvaobj<-fsva(v$E,mod1,svseq,newdat=NULL)

# names_col_new<-names_col
# names_col_new<-sub(paste("(B\\d_",GR1,")",sep=''),paste(GR1),names_col_new)
# names_col_new<-sub(paste("(B\\d_",GR2,")",sep=''),paste(GR2),names_col_new)
```

# PCA with all factors removed `r nsv``

```{r,warning=FALSE,message=FALSE,fig.height = 7, fig.width=7}
Plot_PCA(fsvaobj$db,samples,names_col,group_size,no_of_batches,size_of_batches,col_vector,paste("PCA with all factors removed (",nsv,")"))
```

```{r,warning=FALSE,message=FALSE,include=FALSE}
mmod1 <- cbind(mod1, svseq$sv)
colnames(mmod1) <- c(levels(factor(samples)), paste("X",1:nsv,sep=""))
v2 <- voom(dge,mmod1,plot=FALSE)
an.error.occured <- TRUE
while(an.error.occured==TRUE){
  ok<-TRUE
  fit<-tryCatch(lmFit(v2,mmod1,method='robust', maxit=9999),
         warning=function(w){

        ok<<-FALSE
         })
  if (!ok)
  {
    nsv <-nsv-1
    svseq <- svaseq(to_sva,mod1,mod0,const=0,n.sv=nsv)
    mmod1 <- cbind(mod1, svseq$sv)
    colnames(mmod1) <- c(levels(factor(samples)), paste("X",1:nsv,sep=""))
    v2 <- voom(dge,mmod1,plot=FALSE)
    next}
  
   else { an.error.occured <- FALSE}
}	
#20 minutes?
newV=NULL
fsvaobj<-fsva(v$E,mod1,svseq,newdat=NULL)
```
# Final PCA with `r nsv` factors removed

```{r,warning=FALSE,message=FALSE,fig.height = 7}
Plot_PCA(fsvaobj$db,samples,names_col,group_size,no_of_batches,size_of_batches,col_vector,paste("PCA with ",nsv," factors removed"))

```


```{r,warning=FALSE,message=FALSE,fig.height = 7,include=FALSE}
contrast <- config$CONTRAST_HiSat
P_val<-config$P_VAL_HiSat
cm <- makeContrasts(contrast,levels=mmod1)
cf <- contrasts.fit(fit, cm)
fe <- eBayes(cf)  ## needed for B in topTable, but we ignore this
tt <- topTable(fe,num=Inf)
pr<-1-propTrueNull(tt$P.Value,method='convest')
cat('Conv.est.',pr,'gene level fit\n')
cfe<- eBayes(cf, proportion=pr)
adj.method='BH'


mart<-useMart(biomart="ENSEMBL_MART_ENSEMBL",
              dataset=config$DATSET_MART)

geneinfo<-getBM(c("external_gene_name","ensembl_gene_id"),"ensembl_gene_id",rownames(hisat.g.filtered),mart)
not_found<-hisat.g.filtered[-which(rownames(hisat.g.filtered)%in%geneinfo[,'ensembl_gene_id']),]
list_not_found <-cbind(NA,rownames(not_found))
colnames(list_not_found)<-c('external_gene_name','ensembl_gene_id')
geneinfo<-rbind(geneinfo,list_not_found)
rownames(geneinfo)<-geneinfo[,2]

res_limma<-topTable(cfe, number=Inf,
                adjust.method=adj.method, coef=colnames(cm),
                #p.value=0.05,
                #lfc=0.5,
                sort.by='none',
                confint=TRUE)
score<-apply(res_limma,1,function(x) (
    if (x[1]>0){
      x[2]}
    else
    {-x[3]}
  ))
res_limma$score<-score
res_limma$gene_name<-geneinfo[rownames(res_limma),'external_gene_name']
res_limma<-res_limma[order(res_limma$score,decreasing=TRUE),]
res_limma2<-res_limma[res_limma$adj.P.Val<P_val,]
write.csv(res_limma[,c(10,9,7,1)],paste(hisat_path,'/Results_limma.csv',sep=''))
write.csv(res_limma2[,c(10,9,7,1)],paste(hisat_path,'/Results_limma_pVal_',P_val,'.csv',sep=''))

```

# Number of differentially expressed genes with adjusted p-value < `r P_val`: `r dim(res_limma2)[1] `

# Section 2 -EdgeR
 
```{r,warning=FALSE,message=FALSE,include=FALSE}

### edgeR
y <- DGEList(counts=hisat.g.filtered, group=samples)
colnames(y) <- colnames(hisat.g.filtered,)
keep <- filterByExpr(y)
summary(keep)
y <- y[keep, , keep.lib.sizes=FALSE]
y <- calcNormFactors(y)
design <- mmod1
y <- estimateDisp(y, design, robust=TRUE)
fit <- glmQLFit(y, design, robust=TRUE)
qlf <- glmQLFTest(fit, contrast=cm)
res_edgeR <- topTags(qlf, Inf)$table #default correction: BH
colnames(res_edgeR)[4]<-'adj.P.Val'
res_edgeR2<- res_edgeR[res_edgeR$adj.P.Val<P_val,]
write.csv(res_edgeR,paste(hisat_path,'/Results_edgeR.csv',sep=''))
write.csv(res_edgeR2,paste(hisat_path,'/Results_edgeR_pVal_',P_val,'.csv',sep=''))
#summary(decideTests(qlf))
```

# Number of differentially expressed genes with adjusted p-value < `r P_val`: `r dim(res_edgeR2)[1] `


# Section 3 -DeSeq
 
```{r,warning=FALSE,message=FALSE,include=FALSE}

colnames(mmod1)<-sub('^X','SV',colnames(mmod1))
coldata <- data.frame(condition=samples, type=rep('single-read', length(samples)))
dds <- DESeqDataSetFromMatrix(countData = round(hisat.g.filtered),
                              colData = coldata,
                              design= ~0+condition)#mmod1)
dds <- estimateSizeFactors(dds)
dds <- estimateDispersions(dds)
dds <- nbinomWaldTest(dds, maxit=500)

rownames(cm)<-colnames(mmod1)
tt <- results(dds, contrast=c("condition",names(cm[which(cm!=0),])[1],names(cm[which(cm!=0),])[2]))
res_DeSeq<- tt[is.finite(tt$ad),]
colnames(res_DeSeq)[6]<-'adj.P.Val'
res_DeSeq2<-res_DeSeq[res_DeSeq$adj.P.Val<P_val,]
write.csv(res_DeSeq,paste(hisat_path,'/Results_DeSeq.csv',sep=''))
write.csv(res_DeSeq,paste(hisat_path,'/Results_DeSeq_pVal_',P_val,'.csv',sep=''))
```

# Number of differentially expressed genes with adjusted p-value < `r P_val`: `r dim(res_DeSeq2)[1] `

# Ven diagram for all three methods

```{r,warning=FALSE,message=FALSE}
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")

ven<-venn.diagram(x=list('Limma'=rownames(res_limma2),'EdgeR'=rownames(res_edgeR2),'DeSeq2'=rownames(res_DeSeq2)),filename=NULL,fill = c("yellow","lightgreen","mediumorchid"))
render.venn.diagram(ven)
```

# Barplots for top 10 terms for given GO term

```{r,warning=FALSE,message=FALSE}

# #####topGO
topDiffGenes <- function (allGene, thr=0.05)
{
  return(allGene < thr);
}


for (method in c('limma','edgeR','DeSeq'))
  {
    for (ont  in c('BP','MF','CC'))
    {
    name<-paste('res_',method,sep="")
    if (dim(get(name))[1]!=0)
    {
    geneList<-get(name)$adj.P.Val
    names(geneList)<-rownames(get(name))
    res_GO <- new("topGOdata", ontology = ont,
                  allGenes =geneList, geneSel = topDiffGenes,
                  nodeSize = 10,
                  annot = annFUN.org, mapping=config$GO_MAPPING,ID="ensembl")
    WTP.resultFisher.pc <- runTest(res_GO, algorithm = "parentchild", statistic = "fisher")
      
    used<-usedGO(res_GO)
    GO.allRes <- GenTable(res_GO, #classicFisher = WTP.resultFisher,
                           parentchildFisher=WTP.resultFisher.pc,orderBy = "parentchildFisher",
                           ranksOf = "parentchildFisher",topNodes=length(used))
    colnames(GO.allRes)[6]<-"adj.P.Val"
    GO.allRes2<-GO.allRes[GO.allRes$adj.P.Val<config$TH_GO_HiSat,]
    if (dim(GO.allRes2)[1]<10){b=dim(GO.allRes2)[1]}else {b=10}
    if (b!=0)
      {
      for (j in c(1:b))
        {
          if (GO.allRes2[j,]$adj.P.Val=="< 1e-30")
            {
              GO.allRes2[j,]$adj.P.Val=1e-30
            }
        }
        
      names_BP<-GO.allRes2$Term[b:1]
      par(mai=c(1,3,1,1))
      barplot(abs(log(as.numeric(GO.allRes2[b:1,]$adj.P.Val))),main=paste('Top',b,ont,'for',method,sep=' '),
              horiz=TRUE, names.arg = names_BP,xlim=c(0,50),las=1,col='darkblue',cex.names = 0.7,xlab="abs(log(p-value))")
      }
    
    write.csv(GO.allRes,paste(hisat_path,'/Results_',method,'_',ont,'.csv',sep=''))
    write.csv(GO.allRes2,paste(hisat_path,'/Results_',method,'_',ont,'_',config$TH_GO_HiSat,'.csv',sep=''))
    assign(paste(method,ont,sep='_'),GO.allRes2)
    }
  } 
}

if (length(res_DeSeq2$adj.P.Val)>0 & length(res_edgeR2$adj.P.Val)>0 & length(res_limma2$adj.P.Val)>0)
{

ven<-venn.diagram(x=list('Limma'=limma_BP$GO.ID,'EdgeR'=edgeR_BP$GO.ID,'DeSeq2'=DeSeq_BP[,1]),
                  filename=NULL,fill = c("yellow","lightgreen","mediumorchid"), main= "Biological Process")
render.venn.diagram(ven)

ven<-venn.diagram(x=list('Limma'=rownames(limma_CC),'EdgeR'=rownames(edgeR_CC),'DeSeq2'=rownames(DeSeq_CC)),filename=NULL,fill = c("yellow","lightgreen","mediumorchid"), main= "Cellular Component")
render.venn.diagram(ven)

ven<-venn.diagram(x=list('Limma'=rownames(limma_MF),'EdgeR'=rownames(edgeR_MF),'DeSeq2'=rownames(DeSeq_MF)),filename=NULL,fill = c("yellow","lightgreen","mediumorchid"), main= "Molecular FUnction")
render.venn.diagram(ven)
}else if (length(res_DeSeq2$adj.P.Val)>0 & length(res_edgeR2$adj.P.Val)>0 & length(res_limma2$adj.P.Val)==0)
{
  
ven<-venn.diagram(x=list('EdgeR'=edgeR_BP$GO.ID,'DeSeq2'=DeSeq_BP[,1]),
                  filename=NULL,fill = c("lightgreen","mediumorchid"), main= "Biological Process")
render.venn.diagram(ven)

ven<-venn.diagram(x=list('EdgeR'=rownames(edgeR_CC),'DeSeq2'=rownames(DeSeq_CC)),filename=NULL,fill = c("lightgreen","mediumorchid"), main= "Cellular Component")
render.venn.diagram(ven)

ven<-venn.diagram(x=list('EdgeR'=rownames(edgeR_MF),'DeSeq2'=rownames(DeSeq_MF)),filename=NULL,fill = c("lightgreen","mediumorchid"), main= "Molecular FUnction")
render.venn.diagram(ven)
}else if (length(res_DeSeq2$adj.P.Val)>0 & length(res_edgeR2$adj.P.Val)==0 & length(res_limma2$adj.P.Val)>0)
{
  
ven<-venn.diagram(x=list('Limma'=limma_BP$GO.ID,'DeSeq2'=DeSeq_BP[,1]),
                  filename=NULL,fill = c("lightgreen","mediumorchid"), main= "Biological Process")
render.venn.diagram(ven)

ven<-venn.diagram(x=list('Limma'=rownames(limma_CC),'DeSeq2'=rownames(DeSeq_CC)),filename=NULL,fill = c("lightgreen","mediumorchid"), main= "Cellular Component")
render.venn.diagram(ven)

ven<-venn.diagram(x=list('Limma'=rownames(limma_MF),'DeSeq2'=rownames(DeSeq_MF)),filename=NULL,fill = c("lightgreen","mediumorchid"), main= "Molecular FUnction")
render.venn.diagram(ven)
}else if (length(res_DeSeq2$adj.P.Val)==0 & length(res_edgeR2$adj.P.Val)>0 & length(res_limma2$adj.P.Val)>0)
{
  ven<-venn.diagram(x=list('Limma'=limma_BP$GO.ID,'EdgeR'=edgeR_BP$GO.ID),
                  filename=NULL,fill = c("lightgreen","mediumorchid"), main= "Biological Process")
render.venn.diagram(ven)

ven<-venn.diagram(x=list('Limma'=rownames(limma_CC),'EdgeR'=rownames(edgeR_CC)),filename=NULL,fill = c("lightgreen","mediumorchid"), main= "Cellular Component")
render.venn.diagram(ven)

ven<-venn.diagram(x=list('Limma'=rownames(limma_MF),'EdgeR'=rownames(edgeR_MF)),filename=NULL,fill = c("lightgreen","mediumorchid"), main= "Molecular FUnction")
render.venn.diagram(ven)
}else {print("Not enough results to compare")}
```