---
title: "Illumina HumanHT-12 v4 microarrays analysis"
output: pdf_document
params:
  inpath: "../configs"
---



```{r,warning=FALSE,message=FALSE,include=FALSE}

packages<-c("knitr","RColorBrewer","VennDiagram","veccompare","ggplot2","yaml","ggrepel","gplots")
package.check <- lapply(
  packages,
  FUN = function(x) {
    if (!require(x, character.only = TRUE)) {
      install.packages(x, dependencies = TRUE,repos = "http://cran.us.r-project.org")
      library(x, character.only = TRUE)
    }
  }
)
if (!require("BiocManager", quietly = TRUE))
   { install.packages("BiocManager",repos = "http://cran.us.r-project.org")
BiocManager::install()}
listOfBiocPackages<-c("topGO","limma","sva","AnnotationDbi","vsn","beadarray","illuminaHumanv4.db","affy")
notInstalled <- which( !listOfBiocPackages %in% rownames(installed.packages()) )
## check there's still something left to install
if( length(notInstalled) ) {
    BiocManager::install(listOfBiocPackages[ notInstalled ])
}
lapply(listOfBiocPackages,library, character.only=TRUE)
lapply(packages,require, character.only=TRUE)

path_configs<-params$inpath
#opts_knit$set(root.dir = path)
knitr::opts_chunk$set(echo=FALSE)
config<-read_yaml(paste(path_configs,'/config.yaml',sep=''),fileEncoding = "UTF-8")
Description<-read.table(config$METAFILE_Illumina,header=TRUE,sep=';', stringsAsFactors=FALSE)
output_path<-paste(config$FINALOUTPUT,config$PROJECT,'Microarray_results',sep='/')
dir.create(file.path(output_path), showWarnings = FALSE)

```

```{r,warning=FALSE,message=FALSE}
#common
idats = paste(Description$Path,Description$FIleName,sep='/')
condition = Description$Condition
nnames = paste(condition,Description$Repetition,Description$Batch, sep='_')# Healthy2 and Healthy6 are too young
ref.condition = 'Healthy'
names_col<-unique(condition)
bgx = config$BGX_FILE #BGX file downloaded from Illumina site
```

# RAW data

```{r,warning=FALSE,message=FALSE, include=FALSE}
########################################################limma########################################################
data.limma  = read.idat(idats,bgx,annotation="ILMN_Gene")
```

```{r,warning=FALSE,message=FALSE}
dd<-dim(data.limma)[1]
```
Number of probes: `r dd`

```{r,warning=FALSE,message=FALSE}
microarrays<-data.limma$targets$IDATfile # original names
data.limma$targets$IDATfile<-nnames
colnames(data.limma)<-nnames
prop<-propexpr(data.limma) #Estimate the proportion of probes which are expressed 
                     #above the level of the negative controls
```

## Estimated proportion of probes which are expressed above the level of the negative controls

```{r,warning=FALSE,message=FALSE}
kable(prop,col.names = 'Proportion')
```


```{r,warning=FALSE,message=FALSE}

boxplot(log2(data.limma$E[data.limma$genes$Status=="regular", ]),range=0,las=2,xlab="",ylab=expression(log[2](intensity)),main = "Regular probes, not normalized",names=nnames,cex.axis=0.8)
boxplot(log2(data.limma$E[data.limma$genes$Status=="negative", ]),range=0,las=2,xlab=" ",ylab=expression(log[2](intensity)),main = "Negative control probes",names=nnames,cex.axis=0.8)
#plotDensity(log2(data.limma$E[data.limma$genes$Status=="regular", ]), xlab=expression(log[2](intensity)))
 op <- par(cex = 0.9)

plotDensities(log2(data.limma$E[data.limma$genes$Status=="regular",]),legend="topright", main="Density plot for regular probes",cex=0.7)

```

## MA plots for regular probes

```{r, figures-side, fig.show="hold", out.width="50%"}
for (i in 1:length(nnames))
{
plotMA(data.limma$E[data.limma$genes$Status=="regular",],array=i)
abline(h=0,col='red')
}

```

# NEQC normalization

```{r,warning=FALSE,message=FALSE}
data.limma.norm<-neqc(data.limma) # only regular probes are left for further analysis after this normalisation
dd<-dim(data.limma.norm)[1]
colnames(data.limma.norm)<-nnames
```
Number of probes after NEQC normalization: `r dd`
Only regular probes are left

```{r,warning=FALSE,message=FALSE}

boxplot(data.limma.norm$E,range=0,las=2,xlab=" ",ylab=expression(log[2](intensity)),main = "Regular probes, NEQC normalized",names=nnames)
#MDS plot
plotMDS(data.limma.norm$E,labels=nnames, main= "Regular probes, NEQC normalization")

ids<- as.character(rownames(data.limma.norm))
ids2<- unlist (mget(ids,revmap(illuminaHumanv4ARRAYADDRESS),ifnotfound = NA))
qual<- unlist(mget(ids2, illuminaHumanv4PROBEQUALITY,ifnotfound = NA)) #probe quality
rownames(data.limma.norm)<-ids2
```

Number of probes depending on quality 
'No match' or 'Bad' probes will be removed

```{r,warning=FALSE,message=FALSE}
kable(table(qual))
```

```{r,warning=FALSE,message=FALSE}
AveSignal<-rowMeans(data.limma.norm$E)
boxplot(AveSignal~qual,main='Boxplot of probes qualities')
rem <- qual=="No match" | qual =="Bad" # probes with quality "No match" or "Bad" were removed
data.limma.norm.filt <-data.limma.norm[!rem,]
```



```{r,warning=FALSE,message=FALSE}
dd<-dim(data.limma.norm.filt)[1]
```
Number of probes after normalization and filtering `r dd`

```{r,warning=FALSE,message=FALSE}
boxplot(data.limma.norm.filt$E,range=0,las=2,xlab=" ",ylab=expression(log[2](intensity)),main = "Regular probes, NEQC normalized and filtered",names=nnames)
plotMDS(data.limma.norm.filt$E,labels=nnames, main= "Regular probes, NEQC normalized and filtered")
#clustering
IQR<- apply(data.limma.norm.filt$E,1,IQR,na.rm=TRUE)
topVar<-order(IQR, decreasing=TRUE)[1:100]
d<-dist(t(data.limma.norm.filt$E[topVar,]))
plot(hclust(d),main="Regular probes, NEQC normalized and filtered")
#heatmap
heatmap.2(data.limma.norm.filt$E[topVar,],main="Regular probes,\n NEQC normalized and filtered",key=FALSE,scale="row",cex.main=0.3,density="none",trace="none",dendrogram='column',cexCol=0.8)
```

```{r,warning=FALSE,message=FALSE,include=FALSE}
#DE
condition.fac<-as.factor(condition)
design<-model.matrix(~0+condition.fac)
colnames(design)<-levels(condition.fac)

#Sva
mod1 <- model.matrix(~0+condition.fac)
mod0 <- model.matrix(~1,data=condition.fac)
colnames(mod1) <- levels(condition.fac)
rownames(mod1) <- colnames(data.limma.norm.filt)
to_sva <- 2^data.limma.norm.filt$E
nsv<-num.sv(as.matrix(data.limma$E),mod1,method="leek") 
an.error.occured <- TRUE
while(an.error.occured==TRUE){
   svaa<- try(sva(to_sva,mod1,mod0,n.sv=nsv),TRUE)
   if(isTRUE(class(svaa)=="try-error")) { 
     nsv <-nsv-1
     next
      } 
   else { an.error.occured <- FALSE}
}
newV=NULL
fsvaobj<-fsva(data.limma.norm.filt$E,mod1,svaa,newdat=NULL) # for prediciton only
```

## Plots for all possible factors removed

```{r,warning=FALSE,message=FALSE}
#######################with all factors removed#######################
#clustering
IQR<- apply(fsvaobj$db,1,IQR,na.rm=TRUE)
topVar<-order(IQR, decreasing=TRUE)[1:100]
d<-dist(t(fsvaobj$db[topVar,]))
plot(hclust(d),main="Regular probes, NEQC normalized and filtered\n + SVA all factors removed")
#heatmap
heatmap.2(fsvaobj$db[topVar,],main="Regular probes, NEQC normalized\n and filtered + SVA all factors removed",key=FALSE,scale="row",cex.main=0.3,density="none",trace="none",dendrogram='column',cexCol=0.8)
#MDS plot
plotMDS(fsvaobj$db,labels=nnames,main="Regular probes, NEQC normalized\n and filtered + SVA all factors removed")
#######################with all factors removed#######################
```

```{r,warning=FALSE,message=FALSE,include=FALSE}
mmod1 <- cbind(mod1, svaa$sv)
colnames(mmod1) <- c(levels(condition.fac), paste("X",1:nsv,sep=""))
an.error.occured <- TRUE
while(an.error.occured==TRUE){
  ok<-TRUE
  fit<-tryCatch(lmFit(data.limma.norm.filt$E,mmod1,method='robust', maxit=9999),
         warning=function(w){

        ok<<-FALSE
         })
  if (!ok)
  {
    nsv <-nsv-1
    svaa <- sva(to_sva,mod1,mod0,n.sv=nsv)
    mmod1 <- cbind(mod1, svaa$sv)
    colnames(mmod1) <- c(levels(condition.fac), paste("X",1:nsv,sep=""))
    next}
   else { an.error.occured <- FALSE}
}	
#20 minutes?
newV=NULL
fsvaobj<-fsva(data.limma.norm.filt$E,mod1,svaa,newdat=NULL)
contrast <- config$CONTRAST
P_val<-config$P_VAL_micro
cm <- makeContrasts(contrast,levels=mmod1)
contr.fit<-eBayes(contrasts.fit(fit,cm))
tt <- topTable(contr.fit,num=Inf)
pr<-1-propTrueNull(tt$P.Value,method='convest')
cat('Conv.est.',pr,'gene level fit\n') #0.99
cfe<- eBayes(contrasts.fit(fit,cm), proportion=pr)
adj.method='BH'
options(scipen=999)
res.limma.neqc<-topTable(cfe, number=Inf,
                adjust.method=adj.method, coef=1,
                #p.value=0.05,
                #lfc=0.5,
                sort.by='none',
                confint=TRUE)

score<-apply(res.limma.neqc,1,function(x) (
    if (x[1]>0){
      x[2]}
    else
    {-x[3]}
  ))


res.limma.neqc$score<-score
res.limma.neqc<-res.limma.neqc[order(res.limma.neqc$score,decreasing=TRUE),]
res.limma.neqc2<-res.limma.neqc[res.limma.neqc$adj.P.Val<P_val,]
write.csv(res.limma.neqc,paste(output_path,'/Results_NEQC.csv',sep=''))
write.csv(res.limma.neqc2,paste(output_path,'/Results_NEQC_pVal_',P_val,'.csv',sep=''))
```

## Final plots with `r nsv` factors removed

```{r,warning=FALSE,message=FALSE}
#######################with all factors removed#######################
#clustering
IQR<- apply(fsvaobj$db,1,IQR,na.rm=TRUE)
topVar<-order(IQR, decreasing=TRUE)[1:100]
d<-dist(t(fsvaobj$db[topVar,]))
plot(hclust(d),main=paste("Regular probes, NEQC normalized and filtered\n with ",nsv,' factors',sep=''))
#heatmap
heatmap.2(fsvaobj$db[topVar,],main=paste("Regular probes, NEQC normalized\n and filtered with ",nsv,' factors',sep=''),key=FALSE,scale="row",cex.main=0.2,density="none",trace="none",dendrogram='column',cexCol=0.8)
#MDS plot
plotMDS(fsvaobj$db,labels=nnames,main=paste("Regular probes, NEQC normalized\n and filtered with ",nsv,' factors',sep=''))

```

# NEQC: Number of differentially expressed genes with adjusted p-value < `r P_val`: `r dim(res.limma.neqc2)[1] `

# VSN normalization

```{r,warning=FALSE,message=FALSE,include=FALSE}
########################################################BeadArray########################################################
data.bead= readIdatFiles(idatFiles = idats) #read idat files
pData(data.bead)[,1] = nnames #change default name to user provided names
rownames(pData(data.bead)) = nnames
colnames(exprs(data.bead)) = nnames
pData(data.bead)[,"sampleFac"] = condition #add conditions
pData(data.bead)[,"sampleFac"] = as.factor(pData(data.bead)[,"sampleFac"]) #change into factors
pData(data.bead)[,2] = relevel(pData(data.bead)[,2],ref = ref.condition) #relevel the factors
```

```{r,warning=FALSE,message=FALSE}
bead.eset = normaliseIllumina(BSData = data.bead,method = "vsn",transform = "none") # tronasform to none because vsn transforms it
eset.exprs = as.data.frame(exprs(bead.eset)) #get exprs
names(eset.exprs) = nnames

is.na(eset.exprs) = do.call(cbind,lapply(eset.exprs, is.infinite)) #remove inf values if any.
eset.exprs = as.matrix(eset.exprs[complete.cases(eset.exprs),])
phenoData = new(Class = 'AnnotatedDataFrame',data = pData(data.bead)) # create new pData

eset = ExpressionSet(assayData = as.matrix(eset.exprs),phenoData = phenoData,annotation = 'Humanv4') #create new expressionSet object
eset = addFeatureData(eset,toAdd = c("SYMBOL", "PROBEQUALITY", "CODINGZONE", "PROBESEQUENCE", "GENOMICLOCATION")) #add other features from IlluminaV4 pacakge.
exprs.df = cbind(exprs(eset),as(eset@featureData,Class = 'data.frame'))
exprs.df = exprs.df[,-grep(pattern = 'Row.names',x = colnames(exprs.df))]
no<-which(featureData(eset)$PROBEQUALITY=="Bad" | featureData(eset)$PROBEQUALITY=="No match")
data.bead.filtered.vsn<-eset[-no,] # filetr out "Bad" or "No match" probes
```

```{r,warning=FALSE,message=FALSE}
dd<-dim(data.bead.filtered.vsn)[1]
```

Number of probes after filtering and normalization: `r dd`

```{r,warning=FALSE,message=FALSE}
boxplot(exprs(data.bead.filtered.vsn),range=0,las=2,xlab=" ",ylab=expression(log[2](intensity)),main = "Regular probes, VSN normalized and filtered",names=nnames)
plotMDS(exprs(data.bead.filtered.vsn),labels=nnames, main= "Regular probes, VSN normalized and filtered")
#clustering
IQR<- apply(exprs(data.bead.filtered.vsn),1,IQR,na.rm=TRUE)
topVar<-order(IQR, decreasing=TRUE)[1:100]
d<-dist(t(exprs(data.bead.filtered.vsn)[topVar,]))
plot(hclust(d),main="Regular probes, VSN normalized and filtered")
#heatmap
heatmap.2(exprs(data.bead.filtered.vsn)[topVar,],main="Regular probes,\n VSN normalized and filtered",key=FALSE,scale="row",cex.main=0.3,density="none",trace="none",dendrogram='column',cexCol=0.8)
```
```{r,warning=FALSE,message=FALSE, include=FALSE}

#DE
condition.fac<-as.factor(condition)
design<-model.matrix(~0+condition.fac)
colnames(design)<-levels(condition.fac)

#Sva
mod1 <- model.matrix(~0+condition.fac)
mod0 <- model.matrix(~1,data=condition.fac)
colnames(mod1) <- levels(condition.fac)
rownames(mod1) <- colnames(data.bead.filtered.vsn)
to_sva <- 2^(exprs(data.bead.filtered.vsn))
nsv<-num.sv(exprs(data.bead.filtered.vsn),mod1,method="leek") 
an.error.occured <- TRUE
while(an.error.occured==TRUE){
   svaa<- try(sva(to_sva,mod1,mod0,n.sv=nsv),TRUE)
   if(isTRUE(class(svaa)=="try-error")) { 
     nsv <-nsv-1
     next
      } 
   else { an.error.occured <- FALSE}
}
newV=NULL
fsvaobj<-fsva(exprs(data.bead.filtered.vsn),mod1,svaa,newdat=NULL) # for prediciton only
```

## Plots for all possible factors removed

```{r,warning=FALSE,message=FALSE}
#######################with all factors removed#######################
#clustering
IQR<- apply(fsvaobj$db,1,IQR,na.rm=TRUE)
topVar<-order(IQR, decreasing=TRUE)[1:100]
d<-dist(t(fsvaobj$db[topVar,]))
plot(hclust(d),main="Regular probes, VSN normalized and filtered\n + SVA all factors removed")
#heatmap
heatmap.2(fsvaobj$db[topVar,],main="Regular probes, VSN normalized\n and filtered + SVA all factors removed",key=FALSE,scale="row",cex.main=0.3,density="none",trace="none",dendrogram='column',cexCol=0.8)
#MDS plot
plotMDS(fsvaobj$db,labels=nnames,main="Regular probes, VSN normalized\n and filtered + SVA all factors removed")
#######################with all factors removed#######################
```

```{r,warning=FALSE,message=FALSE,include=FALSE}
mmod1 <- cbind(mod1, svaa$sv)
colnames(mmod1) <- c(levels(condition.fac), paste("X",1:nsv,sep=""))
an.error.occured <- TRUE
while(an.error.occured==TRUE){
  ok<-TRUE
  fit<-tryCatch(lmFit(exprs(data.bead.filtered.vsn),mmod1,method='robust', maxit=9999),
         warning=function(w){

        ok<<-FALSE
         })
  if (!ok)
  {
    nsv <-nsv-1
    svaa <- sva(to_sva,mod1,mod0,n.sv=nsv)
    mmod1 <- cbind(mod1, svaa$sv)
    colnames(mmod1) <- c(levels(condition.fac), paste("X",1:nsv,sep=""))
    next}
   else { an.error.occured <- FALSE}
}	
#20 minutes?
newV=NULL
fsvaobj<-fsva(exprs(data.bead.filtered.vsn),mod1,svaa,newdat=NULL)
contrast <- config$CONTRAST
P_val<-config$P_VAL_micro
cm <- makeContrasts(contrast,levels=mmod1)
contr.fit<-eBayes(contrasts.fit(fit,cm))
tt <- topTable(contr.fit,num=Inf)
pr<-1-propTrueNull(tt$P.Value,method='convest')
cat('Conv.est.',pr,'gene level fit\n') #0.99
cfe<- eBayes(contrasts.fit(fit,cm), proportion=pr)
adj.method='BH'
options(scipen=999)
res.bead.vsn<-topTable(cfe, number=Inf,
                adjust.method=adj.method, coef=1,
                #p.value=0.05,
                #lfc=0.5,
                sort.by='none',
                confint=TRUE)

score<-apply(res.bead.vsn,1,function(x) (
    if (x[1]>0){
      x[2]}
    else
    {-x[3]}
  ))


res.bead.vsn$score<-score
res.bead.vsn<-res.bead.vsn[order(res.bead.vsn$score,decreasing=TRUE),]
res.bead.vsn2<-res.bead.vsn[res.bead.vsn$adj.P.Val<P_val,]
write.csv(res.bead.vsn,paste(output_path,'/Results_VSN.csv',sep=''))
write.csv(res.bead.vsn2,paste(output_path,'/Results_VSN_pVal_',P_val,'.csv',sep=''))
```

## Final plots with `r nsv` factors removed

```{r,warning=FALSE,message=FALSE}
#######################with all factors removed#######################
#clustering
IQR<- apply(fsvaobj$db,1,IQR,na.rm=TRUE)
topVar<-order(IQR, decreasing=TRUE)[1:100]
d<-dist(t(fsvaobj$db[topVar,]))
plot(hclust(d),main=paste("Regular probes, VSN normalized and filtered\n with ",nsv,' factors',sep=''))
#heatmap
heatmap.2(fsvaobj$db[topVar,],main=paste("Regular probes, VSN normalized\n and filtered with ",nsv,' factors',sep=''),key=FALSE,scale="row",cex.main=0.2,density="none",trace="none",dendrogram='column',cexCol=0.8)
#MDS plot
plotMDS(fsvaobj$db,labels=nnames,main=paste("Regular probes, VSN normalized\n and filtered with ",nsv,' factors',sep=''))

```

# VSN: Number of differentially expressed genes with adjusted p-value < `r P_val`: `r dim(res.bead.vsn2)[1] `

# Quantile normalization

```{r,warning=FALSE,message=FALSE}
bead.eset.quantile = normaliseIllumina(BSData = data.bead,method = "quantile",transform = "log2") 
eset.exprs.quantile  = as.data.frame(exprs(bead.eset.quantile)) #get exprs
names(eset.exprs.quantile) = nnames

is.na(eset.exprs.quantile) = do.call(cbind,lapply(eset.exprs.quantile, is.infinite)) #remove inf values if any.
eset.exprs.quantile = as.matrix(eset.exprs.quantile[complete.cases(eset.exprs.quantile),])

phenoData.quantile = new(Class = 'AnnotatedDataFrame',data = pData(data.bead)) # create new pData

eset.quantile = ExpressionSet(assayData = as.matrix(eset.exprs.quantile),phenoData = phenoData.quantile,annotation = 'Humanv4') #create new expressionSet object
eset.quantile = addFeatureData(eset.quantile,toAdd = c("SYMBOL", "PROBEQUALITY", "PROBESEQUENCE", "GENOMICLOCATION")) #add other features from IlluminaV4 pacakge.
exprs.df.quantile = cbind(exprs(eset.quantile),as(eset.quantile@featureData,Class = 'data.frame'))
exprs.df.quantile = exprs.df.quantile[,-grep(pattern = 'Row.names',x = colnames(exprs.df.quantile))]
no<-which(featureData(eset.quantile)$PROBEQUALITY=="Bad" | featureData(eset.quantile)$PROBEQUALITY=="No match")
data.bead.filtered.quantile<-eset.quantile[-no,] 
```

```{r,warning=FALSE,message=FALSE}
dd<-dim(data.bead.filtered.quantile)[1]
```

Number of probes after filtering and normalization: `r dd`

```{r,warning=FALSE,message=FALSE}
boxplot(exprs(data.bead.filtered.quantile),range=0,las=2,xlab=" ",ylab=expression(log[2](intensity)),main = "Regular probes, Quantile normalized and filtered",names=nnames)
plotMDS(exprs(data.bead.filtered.quantile),labels=nnames, main= "Regular probes, Quantile normalized and filtered")
#clustering
IQR<- apply(exprs(data.bead.filtered.quantile),1,IQR,na.rm=TRUE)
topVar<-order(IQR, decreasing=TRUE)[1:100]
d<-dist(t(exprs(data.bead.filtered.quantile)[topVar,]))
plot(hclust(d),main="Regular probes, VSN normalized and filtered")
#heatmap
heatmap.2(exprs(data.bead.filtered.quantile)[topVar,],main="Regular probes,\n Quantile normalized and filtered",key=FALSE,scale="row",cex.main=0.3,density="none",trace="none",dendrogram='column',cexCol=0.8)
```
```{r,warning=FALSE,message=FALSE, include=FALSE}

#DE
condition.fac<-as.factor(condition)
design<-model.matrix(~0+condition.fac)
colnames(design)<-levels(condition.fac)

#Sva
mod1 <- model.matrix(~0+condition.fac)
mod0 <- model.matrix(~1,data=condition.fac)
colnames(mod1) <- levels(condition.fac)
rownames(mod1) <- colnames(data.bead.filtered.quantile)
to_sva <- 2^(exprs(data.bead.filtered.quantile))
nsv<-num.sv(exprs(data.bead.filtered.quantile),mod1,method="leek") 
an.error.occured <- TRUE
while(an.error.occured==TRUE){
   svaa<- try(sva(to_sva,mod1,mod0,n.sv=nsv),TRUE)
   if(isTRUE(class(svaa)=="try-error")) { 
     nsv <-nsv-1
     next
      } 
   else { an.error.occured <- FALSE}
}
newV=NULL
fsvaobj<-fsva(exprs(data.bead.filtered.quantile),mod1,svaa,newdat=NULL) # for prediciton only
```

## Plots for all possible factors removed

```{r,warning=FALSE,message=FALSE}
#######################with all factors removed#######################
#clustering
IQR<- apply(fsvaobj$db,1,IQR,na.rm=TRUE)
topVar<-order(IQR, decreasing=TRUE)[1:100]
d<-dist(t(fsvaobj$db[topVar,]))
plot(hclust(d),main="Regular probes, Quantile normalized and filtered\n + SVA all factors removed")
#heatmap
heatmap.2(fsvaobj$db[topVar,],main="Regular probes, Quantile normalized\n and filtered + SVA all factors removed",key=FALSE,scale="row",cex.main=0.3,density="none",trace="none",dendrogram='column',cexCol=0.8)
#MDS plot
plotMDS(fsvaobj$db,labels=nnames,main="Regular probes, Quantile normalized\n and filtered + SVA all factors removed")
#######################with all factors removed#######################
```

```{r,warning=FALSE,message=FALSE,include=FALSE}
mmod1 <- cbind(mod1, svaa$sv)
colnames(mmod1) <- c(levels(condition.fac), paste("X",1:nsv,sep=""))
an.error.occured <- TRUE
while(an.error.occured==TRUE){
  ok<-TRUE
  fit<-tryCatch(lmFit(exprs(data.bead.filtered.quantile),mmod1,method='robust', maxit=9999),
         warning=function(w){

        ok<<-FALSE
         })
  if (!ok)
  {
    nsv <-nsv-1
    svaa <- sva(to_sva,mod1,mod0,n.sv=nsv)
    mmod1 <- cbind(mod1, svaa$sv)
    colnames(mmod1) <- c(levels(condition.fac), paste("X",1:nsv,sep=""))
    next}
   else { an.error.occured <- FALSE}
}	
#20 minutes?
newV=NULL
fsvaobj<-fsva(exprs(data.bead.filtered.quantile),mod1,svaa,newdat=NULL)
contrast <- config$CONTRAST
P_val<-config$P_VAL_micro
cm <- makeContrasts(contrast,levels=mmod1)
contr.fit<-eBayes(contrasts.fit(fit,cm))
tt <- topTable(contr.fit,num=Inf)
pr<-1-propTrueNull(tt$P.Value,method='convest')
cat('Conv.est.',pr,'gene level fit\n') #0.99
cfe<- eBayes(contrasts.fit(fit,cm), proportion=pr)
adj.method='BH'
options(scipen=999)
res.bead.quantile<-topTable(cfe, number=Inf,
                adjust.method=adj.method, coef=1,
                #p.value=0.05,
                #lfc=0.5,
                sort.by='none',
                confint=TRUE)

score<-apply(res.bead.quantile,1,function(x) (
    if (x[1]>0){
      x[2]}
    else
    {-x[3]}
  ))


res.bead.quantile$score<-score
res.bead.quantile<-res.bead.quantile[order(res.bead.quantile$score,decreasing=TRUE),]
res.bead.quantile2<-res.bead.quantile[res.bead.quantile$adj.P.Val<P_val,]
write.csv(res.bead.quantile,paste(output_path,'/Results_Quantile.csv',sep=''))
write.csv(res.bead.quantile2,paste(output_path,'/Results_Quantile_pVal_',P_val,'.csv',sep=''))
```

## Final plots with `r nsv` factors removed

```{r,warning=FALSE,message=FALSE}
#######################with all factors removed#######################
#clustering
IQR<- apply(fsvaobj$db,1,IQR,na.rm=TRUE)
topVar<-order(IQR, decreasing=TRUE)[1:100]
d<-dist(t(fsvaobj$db[topVar,]))
plot(hclust(d),main=paste("Regular probes, Quantile normalized and filtered\n with ",nsv,' factors',sep=''))
#heatmap
heatmap.2(fsvaobj$db[topVar,],main=paste("Regular probes, Quantile normalized\n and filtered with ",nsv,' factors',sep=''),key=FALSE,scale="row",cex.main=0.2,density="none",trace="none",dendrogram='column',cexCol=0.8)
#MDS plot
plotMDS(fsvaobj$db,labels=nnames,main=paste("Regular probes, Quantile normalized\n and filtered with ",nsv,' factors',sep=''))

```

# Quantile: Number of differentially expressed genes with adjusted p-value < `r P_val`: `r dim(res.bead.vsn2)[1] `

# Venn diagram for all three methods

```{r,warning=FALSE,message=FALSE,include=FALSE}

futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")

ven<-venn.diagram(x=list('NEQC'=rownames(res.limma.neqc2),'VSN'=rownames(res.bead.vsn2),'Quantile'=rownames(res.bead.quantile2)),filename=NULL,fill = c("yellow","lightgreen","mediumorchid"))
```
```{r,warning=FALSE,message=FALSE}
render.venn.diagram(ven)
```
\newpage

# Barplots for top 10 terms for given GO term


```{r,warning=FALSE,message=FALSE, fig.show="hold", out.width="50%"}

# #####topGO
topDiffGenes <- function (allGene, thr=config$TH_GO_micro)
{
  return(allGene < thr);
}


for (method in c('limma.neqc','bead.vsn','bead.quantile'))
  {
    for (ont  in c('BP','MF','CC'))
    {
    name<-paste('res.',method,sep="")
    if (dim(get(name))[1]!=0)
    {
    geneList<-get(name)$adj.P.Val
    names(geneList)<-rownames(get(name))
    res_GO <- new("topGOdata",
              description = paste( method,sep=''), ontology = ont,
              allGenes =geneList, geneSel = topDiffGenes,
              nodeSize = 10,
              annot = annFUN.db,
              affyLib="illuminaHumanv4")

    WTP.resultFisher.pc <- runTest(res_GO, algorithm = "parentchild", statistic = "fisher")

    used<-usedGO(res_GO)
    GO.allRes <- GenTable(res_GO, #classicFisher = WTP.resultFisher,
                           parentchildFisher=WTP.resultFisher.pc,orderBy = "parentchildFisher",
                           ranksOf = "parentchildFisher",topNodes=length(used))
    colnames(GO.allRes)[6]<-"adj.P.Val"
    GO.allRes2<-GO.allRes[GO.allRes$adj.P.Val<config$TH_GO_HiSat,]
      # p.val.FP <- score( WTP.resultFisher.pc)
      # lnames <- names(p.val.FP)[p.val.FP < 0.01]
      # lgenes <- get_anno_genes(lnames, database = 'Mus.musculus',genes=data$gene_name_proper)
      if (dim(GO.allRes2)[1]<10){b=dim(GO.allRes2)[1]
      }else {b=10}
      if (b!=0)
      {
      for (j in c(1:b))
      {
        if (GO.allRes2[j,]$adj.P.Val=="< 1e-30")
        {
          GO.allRes2[j,]$adj.P.Val=1e-30
        }
      }
        
      names_BP<-GO.allRes2$Term[b:1]
      par(mai=c(1,3,1,1))
      barplot(abs(log(as.numeric(GO.allRes2[b:1,]$adj.P.Val))),main=paste('Top',b,ont,'for',method,sep=' '),horiz=TRUE, names.arg = names_BP,xlim=c(0,50),las=1,col='darkblue',cex.names = 0.7,xlab="abs(log(p-value))")
      }
    write.csv(GO.allRes,paste(output_path,'/Results_',method,'_',ont,'.csv',sep=''))
    write.csv(GO.allRes2,paste(output_path,'/Results_',method,'_',ont,'_',config$TH_GO_micro,'.csv',sep=''))
    assign(paste(method,ont,sep='_'),GO.allRes2)
    }
  } 
}

```
\newpage

# Venn diagrams for all three methods

```{r,warning=FALSE,message=FALSE,include=FALSE}

futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")

ven<-venn.diagram(x=list('NEQC'=limma.neqc_BP$GO.ID,'VSN'=bead.vsn_BP$GO.ID,'Quantile'=bead.quantile_BP$GO.ID),
                  filename=NULL,fill = c("yellow","lightgreen","mediumorchid"), main= "Biological Process")
```
```{r,warning=FALSE,message=FALSE}

render.venn.diagram(ven)
```
```{r,warning=FALSE,message=FALSE,include=FALSE}

ven<-venn.diagram(x=list('NEQC'=limma.neqc_CC$GO.ID,'VSN'=bead.vsn_CC$GO.ID,'Quantile'=bead.quantile_CC$GO.ID),filename=NULL,fill = c("yellow","lightgreen","mediumorchid"), main= "Cellular Component")
```
```{r,warning=FALSE,message=FALSE}

render.venn.diagram(ven)
```
```{r,warning=FALSE,message=FALSE,include=FALSE}

ven<-venn.diagram(x=list('NEQC'=limma.neqc_MF$GO.ID,'VSN'=bead.vsn_MF$GO.ID,'Quantile'=bead.quantile_MF$GO.ID),filename=NULL,fill = c("yellow","lightgreen","mediumorchid"), main= "Molecular FUnction")
```
```{r,warning=FALSE,message=FALSE}

render.venn.diagram(ven)
```